ros/src/waypoint_updater/waypoint_updater.py

```diff
@@ -21,36 +22,139 @@
+                final_waypoints.append(self.all_wps.waypoints[i % num_waypoints])
+
+            wps_to_send.waypoints = final_waypoints
+
+            rospy.loginfo("Sending waypoints since current start is different")
+            self.final_waypoints_pub.publish(wps_to_send)
+            self.wp_previous_start = self.wp_current_start
+
+    def check_position(self, pos):
+        num_waypoints = len(self.all_wps.waypoints)
+
+        closest_point = 0
+        max_distance = float('inf')
+        dl = lambda a, b: math.sqrt((a.x-b.x)**2 + (a.y-b.y)**2  + (a.z-b.z)**2)
+
+        for i in range(self.last_closest_waypoint, self.last_closest_waypoint + LOOKAHEAD_WPS):# len(self.all_wps.waypoints)):
+            current_distance = dl(pos, self.all_wps.waypoints[i % num_waypoints].pose.pose.position)
+            if current_distance < max_distance:
+                closest_point = i
+                max_distance = current_distance
+
+        self.last_closest_waypoint = closest_point
+        return closest_point
 
     def waypoints_cb(self, waypoints):
-        # TODO: Implement
-        pass
+        rospy.loginfo("Waypoints received")
+        if (not self.wp_received):
+            self.original_wps = waypoints
+        self.wp_received = True
+        self.all_wps = waypoints
+
+    def current_twist_cb(self, msg):
+        rospy.loginfo("Current twist received")
+        self.current_twist = msg
 
     def traffic_cb(self, msg):
-        # TODO: Callback for /traffic_waypoint message. Implement
-        pass
+        light_wp = msg.data
+
+        rospy.loginfo("Current waypoint %s. traffic waypoints %s", self.wp_current_start, msg.data)
+
+        current_speed = self.get_waypoint_velocity(self.all_wps.waypoints[self.wp_current_start])
+
+        if light_wp >= 0 and current_speed > 0:
+            dist_to_light = self.distance(self.all_wps.waypoints, self.wp_current_start, light_wp) - self.light_wp_offset
+            rospy.logwarn('Braking distance to light: {}'.format(dist_to_light))
+        
+        # at the first moment detecting a red light close by
+        if light_wp >= 0 and not self.stopping:
+            # rospy.loginfo("Traffic waypoints received %s", msg.data)
+            # rospy.loginfo("Target velocity is %s", self.all_wps.waypoints[msg.data].twist.twist.linear.x)
+            # rospy.loginfo("Red light ahead, set velocity to stop")
+
+            wps_to_light = light_wp - self.wp_current_start
+            for i in range(wps_to_light):
+                delta = self.distance(self.all_wps.waypoints, self.wp_current_start, self.wp_current_start + i)
+                if dist_to_light > 0:
+                    new_speed = -1*(current_speed/dist_to_light) * delta + current_speed
+                    self.set_waypoint_velocity(self.all_wps.waypoints, i + self.wp_current_start, new_speed)
+                else:
+                    self.set_waypoint_velocity(self.all_wps.waypoints, j + self.wp_current_start, -1.0)
```

Comment:
I think there is a typo here, j should be i

Reply:
Good catch! I guess that actually never got called yet. Fixed now.